### PR TITLE
Phase 5: Cosine Warm Restarts + SWA — Training Schedule Innovation (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Replace the single-cycle cosine annealing (T_max=180) with CosineAnnealingWarmRestarts and/or Stochastic Weight Averaging (SWA). The current schedule decays the LR monotonically to near-zero — warm restarts periodically reset the LR, which can escape local minima and explore more of the loss landscape. SWA averages weights from multiple points along the trajectory, which provably finds wider optima with better generalization.

**Why this should work:**
- The 8-seed baseline shows p_in has high variance (std=0.58), suggesting the model lands in different local minima depending on the seed. Warm restarts + SWA explore more broadly and average across minima.
- SWA is complementary to EMA — EMA smooths the recent trajectory, SWA averages across restarts. Together they capture both local and global averaging.
- Warm restarts are proven to improve generalization in ImageNet, CIFAR, and regression tasks (Loshchilov & Hutter, ICLR 2017).
- SWA provably converges to wider minima (Izmailov et al., UAI 2018), which correlate with better OOD generalization — directly relevant for p_oodc and p_re.
- Zero architecture change. Compatible with torch.compile. Minimal code change.

**References:**
- Loshchilov & Hutter, "SGDR: Stochastic Gradient Descent with Warm Restarts" (ICLR 2017)
- Izmailov et al., "Averaging Weights Leads to Wider Optima and Better Generalization" (UAI 2018)
- PyTorch: torch.optim.lr_scheduler.CosineAnnealingWarmRestarts, torch.optim.swa_utils

## Instructions

See the first comment for detailed implementation.

## Baseline (8-seed)
val/loss 0.404+/-0.004, p_in 13.33+/-0.58, p_oodc 8.37+/-0.22, p_tan 33.57+/-0.44, p_re 24.58+/-0.13
W&B group: phase5-baseline-8seed

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

All 8 experiments completed successfully (180-minute timeout, ~149-160 epochs for non-SWA, ~149-153 for SWA runs).

### Summary Table

| GPU | Config | W&B Run | val_loss | p_in | p_oodc | p_tan | p_re | Source |
|-----|--------|---------|----------|------|--------|-------|------|--------|
| 0 | wr-T40-m2 | tmuhg2m3 | 0.4082 | 15.16 | 10.04 | 33.97 | 25.52 | ema |
| 1 | wr-T60-m1 | 4yjeyp19 | 0.4149 | 14.72 | 9.42 | 33.64 | 25.47 | ema |
| 2 | wr-T30-m2 | cxoto0ev | 0.4135 | 13.58 | 9.52 | 33.67 | 25.18 | ema |
| 3 | **swa-e120** | ycs4068g | **0.3979** | 13.35 | **8.34** | **32.64** | **24.48** | ema |
| 4 | swa-e100 | m2yvzdj1 | 0.4124 | 12.64 | 8.80 | 34.63 | 24.91 | ema |
| 5 | wr+swa | mazbj68e | 0.4071 | **12.62** | 9.17 | 33.50 | 24.71 | ema |
| 6 | baseline-s42 | h723n624 | 0.4019 | 13.47 | **8.14** | 32.93 | 24.47 | ema |
| 7 | baseline-s43 | yg72l6wb | 0.4094 | 13.88 | 8.72 | 34.31 | 24.63 | ema |

**Baseline reference** (from BASELINE.md): val_loss=0.404+/-0.004, p_in=13.33+/-0.58, p_oodc=8.37+/-0.22, p_tan=33.57+/-0.44, p_re=24.58+/-0.13

### SWA vs EMA Comparison (SWA-enabled runs only)

| Run | EMA val_loss | SWA val_loss | Delta |
|-----|-------------|-------------|-------|
| GPU 3 (swa-e120) | 0.3979 | 0.4009 | SWA worse by 0.003 |
| GPU 4 (swa-e100) | 0.4124 | 0.4154 | SWA worse by 0.003 |
| GPU 5 (wr+swa) | 0.4071 | 0.4116 | SWA worse by 0.005 |

SWA never beat EMA in any run. All saved checkpoints use the EMA source.

### Detailed SWA Surface Pressure Comparison

| Run | | p_in | p_oodc | p_tan | p_re |
|-----|------|------|--------|-------|------|
| GPU 3 EMA | | 13.35 | 8.34 | 32.64 | 24.48 |
| GPU 3 SWA | | 13.46 | 8.53 | 32.65 | 24.55 |
| GPU 4 EMA | | 12.64 | 8.80 | 34.63 | 24.91 |
| GPU 4 SWA | | 12.84 | 9.00 | 34.53 | 25.04 |
| GPU 5 EMA | | 12.62 | 9.17 | 33.50 | 24.71 |
| GPU 5 SWA | | 12.99 | 9.35 | 33.68 | 24.88 |

### Analysis

**Warm Restarts (GPUs 0-2): Did not help.**
- All three warm restart variants performed worse than both baselines (GPUs 6-7).
- GPU 0 (T0=40, Tmult=2): val_loss=0.408, worst of all runs. The periodic LR resets cause destructive interference — the model repeatedly "forgets" fine-grained features it had learned, and with only 2.5 cycles in 160 epochs, it doesn't fully recover.
- GPU 1 (T0=60, Tmult=1): val_loss=0.415, worst overall. Fixed 60-epoch cycles are too long — each restart essentially restarts mid-convergence.
- GPU 2 (T0=30, Tmult=2): val_loss=0.414, slightly better than GPU 1. Shorter initial cycle helps, but still significantly worse than baseline.
- The warm restart schedules produce large LR spikes that hurt surface pressure accuracy across all validation splits.

**SWA (GPUs 3-4): Mixed results.**
- GPU 3 (swa-e120, lr=1e-5): val_loss=0.3979 — **best overall run**, slightly better than the baseline reference (0.404). However, this is the EMA model, not SWA. The SWA model itself (0.4009) is worse than EMA.
- GPU 4 (swa-e100, lr=5e-6): val_loss=0.4124 — worse than baseline. The lower SWA LR and earlier start didn't help.
- Importantly, in both runs, the SWA model consistently underperformed the EMA model by ~0.003. SWA averaging appears too aggressive for this task — it smooths away fine-grained features that EMA preserves.

**Combined wr+swa (GPU 5): Interesting but not better.**
- val_loss=0.4071 — competitive with baseline but doesn't beat GPU 3 or GPU 6.
- Has the best p_in (12.62) across all runs, suggesting warm restarts help the SWA-enabled training explore the in-distribution landscape. But p_oodc (9.17) and p_tan (33.50) are worse than the baseline.

**Baselines (GPUs 6-7): Strong as expected.**
- GPU 6 (seed 42): val_loss=0.4019, p_oodc=8.14 (best across all runs)
- GPU 7 (seed 43): val_loss=0.4094

**Key finding: GPU 3 (swa-e120) achieved the best val_loss (0.3979) of all 8 runs, BUT this came from the EMA model, not SWA.** The SWA training infrastructure may have acted as a mild regularizer (the SWA averaging process slightly modifies the optimization landscape through the additional model averaging), or this may simply be seed variance. The SWA model itself was consistently inferior.

### Commands Used

All runs used the same base command with variant-specific flags:
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=<gpu> nohup python train.py \
  --agent tanjiro --wandb_name "tanjiro/<variant>" \
  --wandb_group "phase5/warm-restarts-swa" \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --disable_pcgrad \
  --pressure_first --pressure_deep \
  <variant-specific flags> \
  > ../logs/p5_gpu<N>_<variant>.log 2>&1 &
```

### Peak Memory

All runs: 36.2 GB VRAM (SWA runs maintain a second copy of the model but this fits within the same allocation).

### Conclusions

1. **Warm restarts hurt performance** for this task. The periodic LR resets are too disruptive for the delicate surface pressure prediction. The model needs stable, monotonically decreasing LR to converge on fine-grained features.
2. **SWA does not beat EMA.** In all three SWA-enabled runs, the SWA model was consistently ~0.003-0.005 worse than the EMA model in val_loss. The exponential moving average (0.999 decay) is already an excellent weight averaging strategy that better preserves recent trajectory information.
3. **No hypothesis variant clearly beat the baseline.** GPU 3's best val_loss (0.3979) is within the baseline confidence interval (0.404 +/- 0.004) and could be seed variance.

### Suggested Follow-ups

1. **Cosine annealing without restarts** — a single cosine schedule (no restarts) with longer total epochs might give the smoothness benefits without the disruptive LR resets.
2. **Higher EMA decay** (e.g., 0.9995 or 0.9999) — if the goal is better weight averaging, tuning EMA decay may be more effective than adding SWA.
3. **Late-stage LR reduction** — instead of SWA's flat LR, try a manual LR step-down at epoch 120+ to squeeze out final convergence.

**Comment by tcapelle** [comment]
ADVISOR: ## Review — SWA-e120 Multi-Seed Validation (Round 2)

**Result: CLOSE — Null. SWA and baseline are statistically indistinguishable.**

4-seed head-to-head:
| Metric | SWA-e120 | Baseline | Delta |
|--------|----------|----------|-------|
| p_in | 13.09±0.24 | 13.50±0.51 | **-3.1%** (SWA wins 4/4 seeds) |
| p_oodc | 8.55±0.21 | 8.46±0.23 | +1.1% (baseline wins) |
| p_tan | 34.07±0.53 | 33.92±0.30 | +0.4% (baseline wins) |
| val/loss | 0.407 | 0.407 | tie |

**Findings:**
1. SWA consistently wins on p_in (-3%) but loses on p_oodc (+1%) and p_tan (+0.4%) — a trade-off, not a clean improvement
2. SWA model always loses to EMA model by ~0.003 — EMA is the better averaging strategy
3. SWA adds ~15s/epoch overhead (153 vs 159 epochs) — throughput cost for no gain
4. R1 p_tan=32.64 finding was seed variance — not reproduced (R2 mean: 34.07)
5. SWA-e120 does reduce p_in variance (0.24 vs 0.51) — more consistent but not better

**Bottom line:** SWA cannot beat our well-tuned EMA. The EMA decay=0.999 with cosine annealing already provides optimal weight averaging for this training regime. Closing.

Excellent validation work, tanjiro. The controlled 4-seed paired comparison is exactly the right way to resolve borderline R1 findings.
